### PR TITLE
dmd.mars: Always produce COFF when targeting Windows

### PR DESCRIPTION
**Pull request by ibuclaw**
Testing whether Azure pipelines will pass.  Optlink is causing issues, [again...](https://github.com/dlang/dmd/pull/12823#issuecomment-875115091)

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Should probably come with a way to revert to OMF if anybody actually wants/needs to use it (maybe `-m32omf`?)

**Comment by thewilsonator** [comment]
see also https://issues.dlang.org/show_bug.cgi?id=18964

**Review by thewilsonator — APPROVED** [review]

**Comment by Geod24** [comment]
```
lib\druntime.lib: Error: Object module `errno_c_32.obj` is 32 bit OMF, but it should be 64 bit MS-Coff
lib\druntime.lib: Error: Object module `minit.obj` is 32 bit OMF, but it should be 64 bit MS-Coff
```

**Comment by ibuclaw** [comment]
> ```
> lib\druntime.lib: Error: Object module `errno_c_32.obj` is 32 bit OMF, but it should be 64 bit MS-Coff
> lib\druntime.lib: Error: Object module `minit.obj` is 32 bit OMF, but it should be 64 bit MS-Coff
> ```

Thanks, just trialing things at the moment (I guess I will probably end with removing the OMF pipelines?).

Tah @thewilsonator - looks like what @ManuEvans has previously done would be a good baseline to begin from.

**Comment by Geod24** [comment]
CC @TurkeyMan

**Comment by ibuclaw** [comment]
Only had a cursory blast through the codebase, I have no means of testing, so just waiting for the CI results.

**Comment by ibuclaw** [comment]
This _can_ be split up into two logical PRs (well, three to remove the dependency on target variables).  No point doing that unless together they all work.

File: src/dmd/link.d
Comment (thewilsonator) [inline src/dmd/link.d:229]: I would prefer if you just negated the condition and left the associated branches as they are to reduce unnecessary diff.

**Comment by kinke** [comment]
> I guess I will probably end with removing the OMF pipelines?

Yes, unless they are still supposed to be tested with `-m32omf`.

**Comment by ibuclaw** [comment]
> > I guess I will probably end with removing the OMF pipelines?
> 
> Yes, unless they are still supposed to be tested with `-m32omf`.

I better add `-m32omf` as a deprecated switch then. ;-)

**Review by WalterBright — COMMENTED** [review]
Changing 18 files and 200 lines just to change a default? DMD is not ready to be packagized. But I digress.

I've hitched my wagon to Microsoft Link before. The troubles are:

1. Versions. Every customer has a different version of that linker. Some would work, some wouldn't. I couldn't just ship someone with a non-working MS-Link a linker that did work.
2. Bugs. Microsoft placed no priority whatsoever on fixing bugs. I was faced with being dead in the water for indeterminate times. Obviously, this was untenable.
3. The compiler didn't work out of the box. I had to tell customers they had to buy some MS dev system to get a linker.
4. When MS-Link didn't work, or crashed, etc., *I* always got blamed.

Having a linker I had control over resolved these issues.

Nothing has changed with those points. It's important that DMD *work out of the box*. I know that we rely on MS-Link for 64 bits, but at least they can get their code working *by default* out of the box.

We can fix Optlink bugs if we need to. I have fixed many. We cannot do jack about MS-Link bugs, except desperately search for some workaround. No attempt has been made even to figure which line of code triggered the Optlink error. Maybe it is trivial, maybe it isn't. That's not really a reason to just abandon it.

I oppose this PR.

**Comment by WalterBright** [comment]
I'd like to redirect this discussion a bit. 32 bit Windows is nearing the end of its useful life. The writing is on the wall. Instead of upending 32 bit Windows support, how about simply changing the default compilation to 64 bits? and leave 32 bit support as it is?

**Comment by dkorpel** [comment]
> Nothing has changed with those points.

What do you think about these points?

- DMD doesn't depend on the MS toolchain anymore, it now ships with the lld linker and MinGW import libs
- dub, the most popular D build tool, has mscoff as the default now since 2019. Just look at the [big list of issues with OPTLINK](https://github.com/dlang/dub/pull/1661) it cited. Notably vide.d, the most popular dub package, doesn't build with OPTLINK because it hits the 32K symbol limit.
- The OMF WinAPI import libraries are outdated. My Windows programs don't build with -m32 because of missing symbols.
- The last OPTLINK bugfix [seems to be from 2015](https://github.com/DigitalMars/optlink/commits/master), but there are still many open [OPTLINK issues](https://github.com/DigitalMars/optlink/issues)
- The OMF format [can be surprising for Windows devs](https://issues.dlang.org/show_bug.cgi?id=18964) that don't even know it exists

I'm sorry to hear MS-Link caused you so much trouble in the past, but let's be real here: currently it's OPTLINK and the outdated OMF target causing all the issues for both users and developers.

**Comment by dkorpel** [comment]
> Instead of upending 32 bit Windows support, how about simply changing the default compilation to 64 bits? and leave 32 bit support as it is?

Sounds like a good idea, but orthogonal to the issue of the maintenance costs that the OMF target brings. It doesn't solve the issue that Iain and Florian are spending their time debugging the Azure pipelines.

**Comment by thewilsonator** [comment]
> No attempt has been made even to figure which line of code triggered the Optlink error. 

Because:
* hardly anyone is familiar with the codebase, GH reports it as something like 2/3 assembly (maybe these are tests? don't know don't care) 1/3 C++. If it organised as badly as DMD no-one will want to become familiar unless it is critical to them, see below
* hardly anyone uses Optlink/OMF intentionally (see the above comment on dub's defaults),
* and no-one depends critically on it. If people run into problems and request assistance on the forums/discord/whatever the first thing anyone will suggest is use `-m32mscoff`

As for your listed points:
1. we ship lld/minGW
2. optlink is far worse, and not developed anymore. meanwhile we have a choice of the MS linker and lld, both of which are actively developed by communities and companies larger and more numerous.
3. we ship lld/minGW _and_ the MS toolchain is free.
4. maybe its damned if you do, damned if you don't. But right now, if MS-link doesn't work its Microsoft's fault. Maybe its low priority and they don't fix it in a timely manner, we have lld as a backup option. If lld doesn't work, it LLVM's fault, fortunately we have proper that work on LLVM and have connections or can fix it themselves. Whereas when optlink fails, you are the only person aside from maybe Rainer that {c,w,s}ould fix it. But you haven't, because the last functional confit was 6 _years_ ago. We can certainly depend on Microsoft/LLVM with a better timeframe than that. 

This falls clearly into a cost/benefit problem of utility/maintainance:

Optlink is good because:
* its fast, 
* we can ship with it
*  and it provides an out of the box experience for Win32 users that {don't, won't can't} use either the MS toolchain or lld/minGW.

Optlink is bad because:
* it has bugs and they don't get fixed in any reasonable timeframe
* it drains development resources dealing with bugs that crop up and we have to work around them because they don't get fixed
* out of the box experience provided is not what people expect when building on windows:
* * e.g. external libraries they want to link to will be in the wrong format, 
* * import libraries are out of date
* * it imposes arbitrary limitations on programs (c.f. dub)

>We can fix Optlink bugs if we need to. 

_You_ can, but you don't. Last fix 6 years ago.

>We cannot do jack about MS-Link bugs,

Maybe, but the same is de facto true for optlink

> except desperately search for some workaround.

We do that for optlink anyways. The above are not true for lld.

**Comment by ibuclaw** [comment]
> Changing 18 files and 200 lines just to change a default? DMD is not ready to be packagized. But I digress.

It doesn't need to change so much.  I just flipped the meaning of a field because it makes it easier to discover what is doing what and where.  It also highlighted issues with some parts of the front-end as well.

**Comment by DoctorNoobingstoneIPresume** [comment]
IMO, the benefits of having, for the same programming language (D) multiple toolchains (and versions of toolchains, and builds of toolhains) (including DMD-built-for-Win32-from-just-DMC) include the ability to test a program in that programming language in multiple environments.

E.g. The binary produced from Example1.d by a certain D toolchain relies on the C/C++ runtime library for the C/C++ toolchain used to build that D toolchain.

(Please correct me if I am wrong.)

This might expose bugs (in Example1.d) which simply do not manifest themselves when building with other D toolchains.

The downside is that this might also expose bugs in that D toolchain (or the underlying C/C++ toolchain) -- as seems to be the case here IIUC.

IMO, the benefit outweights the downside, and we should not abandon a toolchain just because it has some bugs. Tomorrow, bugs might be found in the toolchain we favour today. Plus we would be losing the benefit.

**Comment by DoctorNoobingstoneIPresume** [comment]
While bugs are fixed for a toolchain, is it not possible to have the corresponding automatic checks defined (in "pipelines") as optional ?

E.g. If the build and/or checks for OPTLINK fail, a warning is produced for the PR, but the PR can still be merged ?

IMO, this helps collect information useful for the eventual fixing of the bugs in the toolchain.

**Comment by DoctorNoobingstoneIPresume** [comment]
In order to reduce surprise for users (application developers) who have not used OMF, for DMD for 32-bit Windows, is it possible/acceptable to keep the meaning of command-line flags, but simply emit a warning message in the console when producing OMF files ?

**Comment by thewilsonator** [comment]
> This might expose bugs (in Example1.d) which simply do not manifest themselves when building with other D toolchains.

Yes, but the optlink/OMF toolchain is obsolete or at the very least obsolescent and most of the bugs it finds are bugs/limitations in the toolchains, not in DMD/druntime/phobos. It is a very small fraction of total coverage and it is a continuous burden.

> eventual fixing of the bugs in the toolchain.

Problem is they don't get fixed. The last functional commit to optlink was 6 years ago.

> E.g. If the build and/or checks for OPTLINK fail, a warning is produced for the PR, but the PR can still be merged ?

We should do that anyway.

> emit a warning message in the console when producing OMF files ?

I presume that would happen as part of the deprecation process.

**Comment by DoctorNoobingstoneIPresume** [comment]
One problem with killing OPTLINK and/or OMF and/or DMC support for Github dlang PR's is that it is going to be impossible to bring it back.

And it still has value to people who want a simple way to inter-operate D code (built with DMD) with C++ code (built with DMC).

IMO, there is something... just wrong... to needing a certain C/C++ toolchain (Microsoft, GNU, Clang, MinGW) in order to build and use a D toolchain. We should support all C/C++ toolchains, especially those for which a lot of work has been done over the years, especially by a disproportionately small team.

**Comment by thewilsonator** [comment]
> And it still has value to people who want a simple way to inter-operate D code (built with DMD) with C++ code (built with DMC).

and that is how many people? The point is that this is a cost/benefit problem of utility/maintainance

**Comment by DoctorNoobingstoneIPresume** [comment]
> Problem is they don't get fixed. The last functional commit to optlink was 6 years ago.

I agree that is bad.

(If it seems I have not read the arguments in the other comments made here, I apologize. It is part of my strategy to support my own points-of-view, by ignoring the adverse arguments for which I do not have an answer.) 

But it seems unfair to place the burden on a single person -- "Shame on you, @WalterBright, for being the only one who has ever fixed OPTLINK bugs !".

In the future, we might run into similar situations even with a linker supported by a big Company or by a big Community. Maybe such situations have happened in the past. We might find out that bugs are not quickly fixed even by those larger teams.

The suggestions I dare do are an attempt to split the problem into orthogonal problems (similar to the attempts of others involved in this discussion).

- We can reduce surprise for users not accustomed to OMF.

- We can leave DMC-OPTLINK-OMF support in dlang Github pipelines, and allow merging if those checks fail.

- We can give the maintainer(s) time to fix those bugs, but also information (especially automatically generated information -- i.e. the results of checks for all future dlang/phobos/druntime PR's).

**Comment by DoctorNoobingstoneIPresume** [comment]
> > And it still has value to people who want a simple way to inter-operate D code (built with DMD) with C++ code (built with DMC).
> 
> and that is how many people? The point is that this is a cost/benefit problem of utility/maintainance

In order to reduce cost, when such a check fails, instead of debugging the pipelines, the suggestion is to simply ignore the check.

Maybe also add it to a list of bug proofs, but that takes 5 minutes, and does not really happen that often, I believe.

**Comment by thewilsonator** [comment]
> But it seems unfair to place the burden on a single person

There is no blame given, I want to make that point clear. The problem is that Walter is the only one who knows the codebase and is therefore the only one who can fix it or review a fix.

> The suggestions I dare do are an attempt to split the problem into orthogonal problems

Good.

> We can reduce surprise for users not accustomed to OMF.

Well we should start by making it not the default, which despite the title of this PR is what it does (@ibuclaw).

> We can leave DMC-OPTLINK-OMF support in dlang Github pipelines, and allow merging if those checks fail.

We should also do this.

**Comment by Geod24** [comment]
> The downside is that this might also expose bugs in that D toolchain (or the underlying C/C++ toolchain) -- as seems to be the case here IIUC.

> IMO, the benefit outweights the downside, and we should not abandon a toolchain just because it has some bugs. Tomorrow, bugs might be found in the toolchain we favour today. Plus we would be losing the benefit.

Most people here have been contributors for many years. The person that raised this PR, and the one reporting [issue 18964](https://issues.dlang.org/show_bug.cgi?id=18964)  for well over a decade. The cost analysis have been done many times, and it always came back the same: the downsides far outweigh the potential gain.

I'm saying potential here, because I can't recall a single instance where optlink uncovered a bug in DMD. I really can't. The other way around, on the other side, used to happen multiple times per year (and has significantly decreased since DUB switched to lld by default).

> But it seems unfair to place the burden on a single person

It's not about blame, it's about ownership. optlink is *not* under our control: it lives in a different organization, and very few people here have merge access to it. Even if we did, it's written in plain assembly, has no documentation, no test. Contributing to it has a massive upfront cost, for very little benefit.

**Comment by Geod24** [comment]
> In the future, we might run into similar situations even with a linker supported by a big Company or by a big Community. Maybe such situations have happened in the past. We might find out that bugs are not quickly fixed even by those larger teams.

On one hand, we have optlink bugs, which happens frequently, we have evidence of that. On the other hand, we have hypothetical "some linker" bug. Did we ever hit such case of a non-optlink bug ? Yes, ldc had issues with older versions of `ld.bfd` (you can see it mentioned in each release under "Known issues"). What did they do ? Switched to `gold` by default on Linux. And that's the only instance I can think of. The bug has since been fixed (and I'd wager, was reported before LDC hit it), and using `gold` has other benefits anyways (it's faster).

**Comment by WalterBright** [comment]
> it's about ownership. optlink is not under our control: it lives in a different organization, and very few people here have merge access to it.

If that's blocking anyone, please let me know and I'll fix it.

**Comment by DoctorNoobingstoneIPresume** [comment]
> > it's about ownership. optlink is not under our control: it lives in a different organization, and very few people here have merge access to it.
> 
> If that's blocking anyone, please let me know and I'll fix it.

Does "fix it" mean "fix merge access" or "fix the bug", please ? :'(

**Comment by ibuclaw** [comment]
One phrase that crops up a lot at beerconf is "we need to drain the swamp".  OMF is but one of the many things that need either urgent irrigation or draining, because OOB experience on 32bit is terrible for reasons already given here (and many more, including PIC and PIE support), and telling users to use ms32coff to fix their issues is not acceptable.

Hypothetical linker bugs are being thrown a lot round here as a reason why they are right. These excuses should be struck down by the general discourse.  What if there were no bugs in the linker? People will still have a terrible user experience on 32bit with OMF for the same non-linker related reasons given here by others and we'll still be telling users to use ms32coff anyway.

On Linux and OSX, a lot of stress is emphasised on supporting the toolchain that comes installed on the system. Why then must we the NIH stance on Windows then?

**Comment by DoctorNoobingstoneIPresume** [comment]
> > But it seems unfair to place the burden on a single person
> 
> There is no blame given, I want to make that point clear. The problem is that Walter is the only one who knows the codebase and is therefore the only one who can fix it or review a fix.
> 
> > The suggestions I dare do are an attempt to split the problem into orthogonal problems
> 
> Good.
> 
> > We can reduce surprise for users not accustomed to OMF.
> 
> Well we should start by making it not the default, which despite the title of this PR is what it does (@ibuclaw).
> 
> > We can leave DMC-OPTLINK-OMF support in dlang Github pipelines, and allow merging if those checks fail.
> 
> We should also do this.

Thank you all for answers.

May I ask whether keeping a DMC/OMF/OPTLINK build/check in the pipelines (and making it not-required-for-merging-of-PRs) is acceptable for everyone -- as a compromise to still support that toolchain, but with reduced effort until bugs are fixed ?

**Comment by ibuclaw** [comment]
> Maybe also add it to a list of bug proofs, but that takes 5 minutes, and does not really happen that often, I believe.
> 

I have counted 3 issues in the just as many weeks related to OMF pipeline failures. One of which caused the contributor to spend almost a month going round in circles trying to isolate and work around it. This is not an acceptable use of anyone's time, irrespective of whether they did this on their company time or not.

**Comment by DoctorNoobingstoneIPresume** [comment]
> One phrase that crops up a lot at beerconf is "we need to drain the swamp". OMF is but one of the many things that need either urgent irrigation or draining, because OOB experience on 32bit is terrible for reasons already given here (and many more, including PIC and PIE support), and telling users to use ms32coff to fix their issues is not acceptable.
> 
> Linker bugs (hypothetical or not) are being thrown a lot round here as a reason why they are right. These excuses should be struck down by the general discourse. What if there were no bugs in the linker? People will still have a terrible UX on 32bit with OMF for the same non-linker related reasons given here by others and we'll still be telling users to use ms32coff anyway.

So even with bugs fixed, IIUC, the main problem is that OMF is not used by other toolchains and users are surprised. (Or are there other UX problems with OMF ?)

> On Linux and OSX, a lot of stress is emphasised on supporting the toolchain and ABI that comes installed on the system. Why then must we take the NIH stance on Windows then?

Partial answers:

I believe that the multiple toolchains (and the two object file formats) in Windows have been a driving force behind COM, and that COM has proven useful not just for fixing compatibility (e.g.: it is useful for fine-grained versioning, including in open-source projects, not just for Windows -- 7zip and, until recently anyway, Mozilla).

Also, are we not glad, as users, that there are two competitive open-source C++ toolchains ? And that Visual C++ is also available and competitive ? The opposite of a monopoly (such as Chrome's engine in the world of browsers).

@WalterBright : If OPTLINK is significantly faster than ld, gold and lld, than fixing bugs and making it scalable could encourage both users and contributors. Some projects suffer from long link times and high RAM usage. Currently, I just wish to ask whether it is indeed significantly faster. If so, is that still attributed to it being written in Assembler ? Or is there something in OMF which is not in COFF which makes it faster ?

**Comment by WalterBright** [comment]
> PIC and PIE support

AFAIK that has no relevance to Win32.

> One of which caused the contributor to spend almost a month going round in circles trying to isolate and work around it.

AFAIK it was just assumed to be an optlink bug, because it didn't present with mscoff. But there are other differences between mscoff and omf code generation which could very well be the problem (i.e. a codegen bug). There could be other sources of the problem, too, like the system import libs for OMF were built wrong. I suspect that because the very small number of lines of code in that project were using unusual Windows API calls.

It just seems peculiar to me that adding a hundred lines of code to dmd suddenly causes Optlink failures, when we add hundreds of lines of code all the time to dmd and there are no such failures. So what's different about it? The different Windows API calls.

Another common source of strange errors when changing the toolchain is inadvertent use of undefined behavior, like uninitialized stack variables (though that shouldn't happen with D). I've certainly encountered many such coding bugs in the past, and one reason why supporting multiple tool chains is a good idea is it exposes latent bugs.

It's too easy to dismiss it as Optlink, as the actual problem was never found. Now, if Optlink does produce the Unexpected Termination display, that is definitely an Optlink bug.

**Comment by WalterBright** [comment]
> I have counted 3 issues in the just as many weeks

I filed one of them. Were bugzilla issues filed for the other two?

**Comment by WalterBright** [comment]
> On Linux and OSX, a lot of stress is emphasized on supporting the toolchain and ABI that comes installed on the system. Why then must we take the NIH stance on Windows then?

1. the Microsoft toolchain does not come with Windows.
2. on Linux and OSX, it's trivial for users to apt-get the latest toolchain, so dealing with out-of-date parts of the toolchain is not necessary.

**Comment by thewilsonator** [comment]
> the Microsoft toolchain does not come with Windows.

I would like to stress again that we ship `lld` (LLVM based linker) and `minGW` import libraries on windows. Lack of toolchain is not a problem.

**Comment by ibuclaw** [comment]
> > PIC and PIE support
> 
> AFAIK that has no relevance to Win32.
> 

Fraying 32-bit support in general is blocking meaningful progress on all platforms.  32-bit DMD Linux releases have had the compiler segfaulting for a few years now (unless you are running on an EOL system).  The fix is easy, but PIC bugs in the DMD backend are blocking it.

> > One of which caused the contributor to spend almost a month going round in circles trying to isolate and work around it.
> 
> AFAIK it was just assumed to be an optlink bug, because it didn't present with mscoff. But there are other differences between mscoff and omf code generation which could very well be the problem (i.e. a codegen bug). There could be other sources of the problem, too, like the system import libs for OMF were built wrong. I suspect that because the very small number of lines of code in that project were using unusual Windows API calls.
> 

Another reason to sunset OMF.  With COFF, if a problem occurs in DMD pipelines, you can try compiling the same code with either GDC or LDC to see if the same crash occurs.  If it doesn't, you know the fault is DMD.  Because neither GCC or LLVM supports OMF we have no way to track down, verify, or debug problems.

> It's too easy to dismiss it as Optlink, as the actual problem was never found. Now, if Optlink does produce the Unexpected Termination display, that is definitely an Optlink bug.

People have been complaining [for years](https://github.com/dlang/dub/pull/1661) in bugzilla, on forums, in other D projects repos.  It would be harder to dismiss if this was just a one-off, or a rare occurrence.  But in reality, the backlog of evidence is too damn high, and first-time contributors [continually](https://github.com/dlang/dmd/pull/7099/files) run into heisenbugs that manifest only on DMD 32-bit.

**Comment by DoctorNoobingstoneIPresume** [comment]
> > Maybe also add it to a list of bug proofs, but that takes 5 minutes, and does not really happen that often, I believe.
> 
> I have counted 3 issues in the just as many weeks related to OMF pipeline failures. One of which caused the contributor to spend almost a month going round in circles trying to isolate and work around it. This is not an acceptable use of anyone's time, irrespective of whether they did this on their company time or not.

Can such unacceptable spending of time and effort not be avoided by having the DMC/OPTLINK/OMF check temporarily be marked as not required for merging of checked dlang/druntime/phobos PRs (but still performed by our Github pipelines) ?

**Comment by ibuclaw** [comment]
> Can such unacceptable spending of time and effort not be avoided (or at least reduced) by having the DMC/OPTLINK/OMF check temporarily be marked as not required for merging of checked dlang/druntime/phobos PRs (but still performed by our Github pipelines) ?

We release OMF binaries, so unless that stops...

**Comment by DoctorNoobingstoneIPresume** [comment]
> > it's about ownership. optlink is not under our control: it lives in a different organization, and very few people here have merge access to it.
> 
> If that's blocking anyone, please let me know and I'll fix it.

For me personally, it is my own laziness and lack of ability which is blocking me from learning to work on OPTLINK. If I had been able to fix the bug, I am sure I could convince the maintainer to merge that fix. No, what I really want is support for this toolchain, but at the expense of effort made by others. Why do I want it ? I have tried explaining, but got "How many others want support for DMC/OPTLINK/OMF ?". It is difficult for me to answer that.

**Comment by DoctorNoobingstoneIPresume** [comment]
> > Can such unacceptable spending of time and effort not be avoided (or at least reduced) by having the DMC/OPTLINK/OMF check temporarily be marked as not required for merging of checked dlang/druntime/phobos PRs (but still performed by our Github pipelines) ?
> 
> We release OMF binaries, so unless that stops...

IIUC, you want support for OMF to be eliminated completely, and you have been using the argument of time and effort needed to investigate fails of OMF checks.

I am proposing, as a compromise, from the point-of-view of a non-contributor and of possibly a weird user, that such failures are simply not investigated. I feel this leaves at least some form of support for OMF.

Regarding changes of default target from OMF to COFF, FWMOIW (for whatever my opinion is worth): I would vote against, but if you feel strongly for that, I can live with that just fine.

Why have I dared answer to this conversation ? I felt that the following question ws being asked: "Is anyone still using OPTLINK or OMF [or interoperating with DMC] [or happy for small size of toolchain and independence] [or appreciative of testing with multiple toolchain versions] ?"

So I have answered.

Have I been desconsiderate of other people's time and effort ? Maybe. I apologize if so. Until today, I had not been aware that OPTLINK has bugs. Even so, dropping it feels too harsh to me, especially when a compromise of ignoring that pipeline [until OPTLINK is fixed, if ever] can be made (sad as it may seem to me).

**Comment by DoctorNoobingstoneIPresume** [comment]
> > In the future, we might run into similar situations even with a linker supported by a big Company or by a big Community. Maybe such situations have happened in the past. We might find out that bugs are not quickly fixed even by those larger teams.
> 
> On one hand, we have optlink bugs, which happens frequently, we have evidence of that. On the other hand, we have hypothetical "some linker" bug. Did we ever hit such case of a non-optlink bug ? Yes, ldc had issues with older versions of `ld.bfd` (you can see it mentioned in each release under "Known issues"). What did they do ? Switched to `gold` by default on Linux. And that's the only instance I can think of. The bug has since been fixed (and I'd wager, was reported before LDC hit it), and using `gold` has other benefits anyways (it's faster).

gold is faster than ld, but is it faster than lld or vice-versa ?

**Comment by rikkimax** [comment]
> * and no-one depends critically on it. If people run into problems and request assistance on the forums/discord/whatever the first thing anyone will suggest is use `-m32mscoff`

I can't remember the last time I or anyone else recommended anyone to make a 32bit build on Discord.

Straight to 64bit. 

Ever since dub switched to coff by default for 32bit, I don't remember having to question this at all.

**Comment by TurkeyMan** [comment]
Reading through this discussion invokes an overwhelming and visceral sense of PTSD.
I'll climb back in my hole...

**Comment by ibuclaw** [comment]
> Reading through this discussion invokes an overwhelming and visceral sense of PTSD.
> I'll climb back in my hole...

@TurkeyMan at least optlink is now bowing out and coff is the default for 32bit now too. :-)

**Comment by thewilsonator** [comment]
@TurkeyMan this was closed because I rebased it and that got merged.

**Pull request closed without merging**
